README.md
# cloud_validator

A Validator for the Cloud that runs in currently 3 different ways'
1. Validator Light, for quickly building a couple of VMs
2. Validator Custom, for running custom builds easily defined in yaml files (see below)
3. Validator Heavy, which can only be run when a region is empty, and builds as many VMs as possible.

# Validator Custom

Validator Custom allows custom project setups to be defined in yaml files and then be built with Validator.

## `project`
- `name` - Name of the project. Should be unique. (required)

## `subnets`
- `name` - Subnet name. Should be unique within a project. (required)
- `gateway` - The IP address of the gateway. (required)
- `mask` - The mask for the subnet (required)

## `vpn`
- `local_subnet` - local(cloud cix/alpha) private ntw from the project subnets. (required)
- `customer_subnets` - list of remote(user end) private ntw address(ipv4). (required)
- `ike_authentication` - a must field from static list [md5, sha1, sha-256, sha-384]
- `ike_encryption` - a must field from static list [aes-128-cbc, aes-192-cbc, aes-256-cbc, des-cbc, 3des-cbc]
- `ike_lifetime` - an interger value within range [180, 86400] (optional)
- `ike_dh_groups` - a must field from static list [group1, group2, group5, group19, group20, group24]
- `ike_pre_shared_key` - password(not more than 64 chars) for vpn connection establish. (required)
- `ike_version` - a must field from static list [v1-only, v2-only]
- `ike_mode` - a must field from static list [main, aggressive]
- `ike_public_ip` - floating IP address of the user
- `ipsec_authentication` - a must field from static list [hmac-md5-96, hmac-sha1-96, hmac-sha-256-96, hmac-sha-256-128]
- `ipsec_encryption` - a must field from static list [aes-256-cbc, aes-192-cbc, aes-256-cbc, des-cbc, 3des-cbc, aes-128-gcm, aes-192-gcm, aes-256-gcm]
- `ipsec_lifetime` - an interger value within range [180, 86400] (optional)
- `ipsec_pfs_groups` - a must field from static list [group1, group2, group5, group14, group19, group20, group24]
- `ipsec_establish_time` - a must field from static list [immediately, on-traffic]

## `vms`
- `random` - True if project should contain a certain number of randomly generated VMs. (required)
- `count` - The number of random VMs to be created. (required if random is True)
- `vm_list` - List of predefined VMs

### `vm_list`
- `name` - Name of the VM. (required)
- `dns` - The domain name servers to be used by the project. (required)
- `gateway_subnet` - RFC 1918 address range from subnets that the Gateway for the VM will be defined on. 
- `ip_addresses` The IP address for the VM. Must be within one of the defined subnets. Not required. If not included a random IP within one of the subnets will be assigned
  - `address`: IP Address of VM
  - `nat`: Boolean, only IPs from the gateway subnet can be NATed
- `cpu` - Number of CPU cores to be assigned. (required)
- `ram` - Amount of RAM in GB to be assigned. (required)
- `image_id` - ID of the operating system to be installed. (required)
  - 2 - Windows Server 2012
  - 3 - Windows Server 2016
  - 6 - Ubuntu Server 16.04
  - 10 - CentOS Linux 7
  - 12 - Ubuntu Server 18.04
  - 13 - Windows Server 2019
  - 14 - Manual
  - 15 - Red Hat Server 7.7
  - 16 - Centos Linux 8
  - 17 - Ubuntu Server 20.04
- `replicate` - Number of replicates of a VM there should be within a project. e.g 5 would produce five of the same VM. Not required.
- `storage_type_id` - ID of the storage type (required)
  - 1 is HDD
  - 2 is SSD
- `storage` - List of storage for the VM

#### `storage`
- `name` - Name of the storage. Should be unique within the VM. (required)
- `gb` - Size of the storage in GB. (required)
- `primary` - True if primary storage. Every VM must have one storage marked as primary. (required)


#### `firewall_rules`
- `allow` - True if the traffic matching the rule should be allowed through the firewall. (required)
- `destination` - A Subnet or IP Address representing the destination value for the rule. Use * to represent all. (required)
- `port` - The port to use when checking incoming traffic against this rule. No port is required if the protocol is icmp or any
- `protocol` - The protocol to use when checking incoming traffic against this rule. Options are [icmp, tcp, udp, any]. (required)
- `source` - A Subnet or IP Address representing the source value for the rule. Use * to represent all. (required)


## Example

```yaml
project:
  name: Test-Project

subnets:
  - name: Subnet-One
    gateway: 192.168.123.1
    mask: 24

vpn:
  local_subnet: 192.168.123.1/24
  customer_subnets: [172.16.32.0/24]
  ike_authentication: md5
  ike_encryption: aes-256-cbc
  ike_lifetime: 18000
  ike_dh_groups: group2
  ike_pre_shared_key: test
  ike_version: v1-only
  ike_mode: main
  ike_public_ip: 91.103.1.30
  ipsec_authentication: hmac-md5-96
  ipsec_encryption: aes-256-cbc
  ipsec_lifetime: 18000
  ipsec_pfs_groups: group2
  ipsec_establish_time: immediately

vms:
  random: true
  count: 2
  vm_list:
    - name: VM-ONE
      dns: 91.103.0.1,8.8.8.8
      cpu: 1
      ram: 1
      image_id: 10
      gateway_subnet: 192.168.123.1/24
      ip_addresses:
        - address: 192.168.123.2
          nat: True
      storage_type_id: 1
      storage:
        - name: 'HDD-ONE'
          gb: 50
          primary: true
        - name: 'HDD-TWO'
          gb: 50
          primary: false
    - name: VM-REPLICATE
      dns: 91.103.0.1,8.8.8.8
      cpu: 1
      ram: 1
      image_id: 12
      storage_type_id: 1
      replicate: 2
      storage:
        - name: 'HDD-ONE'
          gb: 50
          primary: true

  firewall_rules:
    - allow: true
      source: '*'
      destination: 192.168.132.2/32
      port: '20-25'
      protocol: 'tcp'
    - allow: true
      source: '*'
      destination: 192.168.132.2/32
      protocol: 'icmp'

```
# Settings

`Region`: where IAAS(Infrastructure As A Service) is physically(VMs, VPNs, VRs) stored on servers and routers.

`COP`: where IAAS is soft stored(database).

`Robot`: Every Region has Robot functionality and has an account in its COP.

`validating Region`: It is the Region you are trying to validate its functionality with this software.

`validating COP` : It is the COP where your validating Region is adopted/registered to.

`User account`: One must have an account registered in validating COP to make API requests.

`Robot account`: validating Region is an address and has an account in its COP. 

- `CLOUDCIX_API_URL` - API url for the validating COP.
- `CLOUDCIX_API_USERNAME` - Username of the user account registered in validating COP.
- `CLOUDCIX_API_PASSWORD` - Password of the above user account.
- `CLOUDCIX_API_KEY` - API key provided by the validating COP to the above user.
- `CLOUDCIX_API_VERSION` - Current CloudCIX api version, which is 2.
- `CLOUDCIX_API_V2_URL` - V2 API url for the validating COP.
- `ROBOT_USERNAME` - Validating Region's Robot account username.
- `ROBOT_PASSWORD` - Validating Region's Robot account password
- `ROBOT_API_KEY` - Validating Region's Robot account API key.


All above `CLOUDCIX_API_` and `ROBOT_` fields are required and supplied in settings.py file to run this software smoothly.
These fields can be received from your PAM Operator.


CHANGELOG.md
# CHANGELOG
Keep track of changes made for each version

## Pre-Open Source
The following features were created before the project was made open-source. Version numbers will be reset to
version 0 for public release.

## 1.0.0
- Initial deployment

dataclasses/data.py
from __future__ import annotations
# stdlib
import os
import random
import yaml
from netaddr import IPAddress, IPNetwork, IPSet
from sys import exit
from typing import Any, Dict, List, Optional
# cloudcix
os.environ['CLOUDCIX_SETTINGS_MODULE'] = 'settings'
from cloudcix import api  # noqa: E402


class Data:
    """
    Class representing the data being sent to the API to create and build a project.
    """

    firewall_rules: List[Dict[str, Any]]
    images: List[Dict[str, Any]]
    project: Dict[str, Any]
    subnets: List[Dict[str, Any]]
    used: List[str]
    vms: List[Dict[str, Any]]
    vpns: List[Optional[Dict[str, Any]]]

    def __init__(self):
        self.firewall_rules = []
        self.images = []
        self.project = {}
        self.subnets = []
        self.used = []
        self.vms = []
        self.vpns = []

    def add_vm(self):
        """
        Add a new VM to the current data.
        """
        image = random.choice(self.images)
        gateway_subnet = random.choice(self.subnets)
        gateway_ip = self.choose_ip(gateway_subnet['address_range'])
        self.used.append(gateway_ip)
        self.vms.append({
            'image_id': f'{image["id"]}',
            'name': f'New-TestVM-{image["display_name"].replace(" ", "-")}',
            'ram': 1,
            'gateway_subnet': gateway_subnet['address_range'],
            'ip_addresses': [
                {
                    'address': gateway_ip,
                    'nat': True,
                },
            ],
            'dns': '8.8.8.8,8.8.4.4',
            'cpu': 1,
            'storage_type_id': 1,
            'storages': [
                {
                    'primary': True,
                    'name': 'NewHDD',
                    'gb': 50,
                },
            ],
        })

    def add_firewall_rule(self):
        """
        Add a new firewall to the current data.
        """
        subnet = random.choice(self.subnets)
        self.firewall_rules.append(
            {
                'allow': True,
                'source': '91.103.3.36',
                'destination': f'{subnet["address_range"]}',
                'protocol': 'any',
                'debug_logging': False,
                'pci_logging': False,
            },
        )

    def retrieve_images(self, region: str, token: str):
        """
        Method to retrieve all the images in a given region
        """
        params = {
            'search[regions__region]': region,
            'search[enabled]': True,
            'order': 'name',
        }
        response = api.IAAS.image.list(token=token, params=params)
        if response.status_code == 200:
            self.images = response.json()['content']
        # 500 error is known and is under investigation, we are ignoring the effect.
        elif response.status_code == 500:
            pass
        else:
            print('\n\033[91m - There was an error while retrieving images for the region. \033[0m')
            exit(1)

    def choose_ip(self, subnet) -> str:
        """
        Method for selecting an IP for a VM
        """
        ip = ''
        if not subnet:
            # Randomly select subnet
            subnets = [subnet['address_range'] for subnet in self.subnets]
            subnet = random.choice(subnets)

        # Create IP Network and choose IP from valid IPs
        network = IPNetwork(subnet)
        blocked_ips = [network.ip, network.broadcast]
        blocked_ips.extend([IPAddress(used) for used in self.used])
        blocked = IPSet(blocked_ips)
        while not ip or ip in blocked:
            ip = network[random.randint(1, len(network) - 2)]
        return str(ip)

    @classmethod
    def validator_light(cls, region: str, name: str, token: str) -> Data:
        """
        Class method to instantiate an instance of the Data class with set values.
        """

        # Create an instance of the Data class and retrieve the images for the region
        data = cls()
        data.retrieve_images(region=region, token=token)

        # Add project data.
        data.project = {
            'region_id': region,
            'name': name,
        }

        # Add project subnets.
        data.subnets = [{
            'address_range': '192.168.123.1/24',
            'name': 'Test',
        }]

        # Add firewall rules
        data.firewall_rules = [{
            'allow': True,
            'source': '*',
            'destination': '192.168.123.1/24',
            'protocol': 'any',
            'debug_logging': False,
            'pci_logging': False,
        }]

        # Add vpn
        data.vpns = [{
            'description': 'Home',
            'local_subnet': '192.168.123.1/24',
            'customer_subnets': ['172.16.32.0/24'],
            'vpn_type': 'site_to_site',
            'ike_authentication': 'md5',
            'ike_encryption': 'aes-256-cbc',
            'ike_lifetime': 18000,
            'ike_dh_groups': 'group2',
            'ike_pre_shared_key': 'test',
            'ike_version': 'v1-only',
            'ike_mode': 'main',
            'ike_public_ip': '91.103.3.36',
            'ipsec_authentication': 'hmac-md5-96',
            'ipsec_encryption': 'aes-256-cbc',
            'ipsec_lifetime': 18000,
            'ipsec_pfs_groups': 'group2',
            'ipsec_establish_time': 'immediately',
        }]

        # Get Subnets and used IPs for use in setting VM IPs
        data.used = [subnet['address_range'].split('/')[0] for subnet in data.subnets]

        all_subnets = data.subnets
        # Add VM for each image retrieved
        for image in data.images:
            ip_addresses = []
            gateway_subnet = random.choice(data.subnets)
            gateway_ip = data.choose_ip(gateway_subnet['address_range'])
            ip_addresses.append({
                'address': gateway_ip,
                'nat': True,
            })
            rest_subnets = list(
                filter(
                    lambda sub: sub['address_range'] != gateway_subnet['address_range'],
                    all_subnets,
                ),
            )
            if image['multiple_ips']:
                for sub in rest_subnets:
                    ip = data.choose_ip(sub['address_range'])
                    data.used.append(ip)
                    ip_addresses.append({
                        'address': ip,
                        'nat': False,
                    })

            data.vms.append({
                'image_id': f'{image["id"]}',
                'name': f'TestVM-{image["display_name"].replace(" ", "-")}',
                'ram': 1,
                'gateway_subnet': gateway_subnet['address_range'],
                'ip_addresses': ip_addresses,
                'dns': '8.8.8.8,8.8.4.4',
                'cpu': 1,
                'storage_type_id': 1,
                'storages': [
                    {
                        'primary': True,
                        'name': 'C',
                        'gb': 50,
                    },
                ],
            })

        return data

    @classmethod
    def validator_custom(cls, region: str, file: str, token: str) -> Data:
        """
        Class method to instantiate an instance of the Data class using a YAML file.
        """

        # Load config from file
        with open(f'configs/{file}', 'r') as config:
            config_loaded = yaml.safe_load(config)

        # Create an instance of the Data class and retrieve the images for the region
        data = cls()
        data.retrieve_images(region=region, token=token)

        # Set project data
        data.project = {
            'region_id': region,
            'name': config_loaded['project']['name'],
        }

        data.firewall_rules = config_loaded['firewall_rules']

        # Add Subnets to project
        for subnet in config_loaded['subnets']:
            data.subnets.append({
                'address_range': f'{subnet["gateway"]}/{subnet["mask"]}',
                'name': subnet['name'],
            })

        # Add VPN
        try:
            data.vpns = config_loaded['vpns']
        except KeyError:
            pass

        # Get Subnets and used IPs for use in setting VM IPs
        data.used = []
        data.used = [subnet['gateway'] for subnet in config_loaded['subnets']]
        all_subnets = data.subnets
        # Set VMs for random
        if config_loaded['vms']['random']:
            hd_sizes = [50, 100, 150, 200, 250]
            count = config_loaded['vms']['count']
            for vm in range(count):
                # Get IP Address and image for VM
                image = random.choice(data.images)
                gateway_subnet = config_loaded['vms']['vm_list'][0]['gateway_subnet']
                gateway_ip = data.choose_ip(gateway_subnet)
                data.used.append(gateway_ip)

                ip_addresses = config_loaded['vms']['vm_list'][0]['ip_addresses']
                if len(ip_addresses) > 1 and image['multiple_ips']:
                    rest_subnets = list(
                        filter(
                            lambda sub: sub['address_range'] != gateway_subnet['address_range'],
                            all_subnets,
                        ),
                    )
                    subnet = random.choice(rest_subnets)
                    ip = data.choose_ip(subnet['address_range'])
                    data.used.append(ip)
                    for ip_addr in ip_addresses:
                        if 'address' not in ip_addr.keys():
                            if ip_addr['nat']:
                                ip_addr['address'] = gateway_ip
                            else:
                                ip_addr['address'] = ip
                else:
                    ip_addresses = [{
                        'address': gateway_ip,
                        'nat': True,
                    }]

                # Add the VM
                data.vms.append({
                    'image_id': f'{image["id"]}',
                    'name': f'RandomVM-{image["display_name"].replace(" ", "-")}-{vm}',
                    'ram': random.randint(1, 2),
                    'gateway_subnet': gateway_subnet,
                    'ip_addresses': ip_addresses,
                    'dns': '8.8.8.8,8.8.4.4',
                    'cpu': random.randint(1, 2),
                    'storage_type_id': 1,
                    'storages': [],
                })

                # Add Storage
                for storage in range(random.randint(1, 2)):
                    data.vms[vm]['storages'].append({
                        'primary': storage == 0,
                        'name': f'TestHD-{storage}',
                        'gb': random.choice(hd_sizes),
                    })

        if not config_loaded['vms'].get('vm_list'):
            return data

        for vm_obj in config_loaded['vms']['vm_list']:
            replicate = 1 if not vm_obj.get('replicate') else vm_obj.get('replicate')
            for rep in range(replicate):

                # Get name and ip address for VM
                rep_name = f'{vm_obj["name"]}-{rep}'
                name = rep_name if replicate > 1 else vm_obj['name']
                gateway_ip = data.choose_ip(vm_obj['gateway_subnet'])
                data.used.append(gateway_ip)
                ip_addresses = vm_obj['ip_addresses']
                for ip_addr in ip_addresses:
                    if 'address' not in ip_addr.keys():
                        if ip_addr['nat']:
                            ip_addr['address'] = gateway_ip
                        else:
                            rest_subnets = list(
                                filter(
                                    lambda sub: sub['address_range'] != vm_obj['gateway_subnet'],
                                    all_subnets,
                                ),
                            )
                            subnet = random.choice(rest_subnets)
                            ip = data.choose_ip(subnet['address_range'])
                            ip_addr['address'] = ip
                            data.used.append(ip)

                # Add VM
                data.vms.append({
                    'image_id': f'{vm_obj["image_id"]}',
                    'name': name,
                    'ram': vm_obj['ram'],
                    'gateway_subnet': vm_obj['gateway_subnet'],
                    'ip_addresses': ip_addresses,
                    'dns': vm_obj['dns'],
                    'cpu': vm_obj['cpu'],
                    'storage_type_id': vm_obj['storage_type_id'],
                    'storages': vm_obj['storages'],
                })

        return data

    @classmethod
    def validator_heavy(
            cls,
            region: str,
            name: str,
            token: str,
            cores: int,
            ram: int,
            storage: int,
            unix: bool,
            storage_type_id: int) -> Data:
        """
        Class method to instantiate an instance of the Data class which will fill the region
        """

        # Create an instance of the Data class and retrieve the images for the region
        data = cls()
        data.retrieve_images(region=region, token=token)
        windows_images = [image for image in data.images if image['answer_file_name'] == 'windows']
        unix_images = [
            image for image in data.images if
            image['answer_file_name'] == 'ubuntu' or image['answer_file_name'] == 'centos'
        ]

        # Check that there are images for the chosen region and type
        # Check that there are images for the chosen region and type
        if unix and len(unix_images) == 0:
            exit(0)
        if not unix and len(windows_images) == 0:
            exit(0)

        # Add project data.
        data.project = {
            'region_id': region,
            'name': name,
        }

        # Add project subnets.
        data.subnets = [{
            'address_range': '192.168.123.1/24',
            'name': 'Test',
        }]

        # Add firewall rules
        data.firewall_rules = [{
            'allow': True,
            'source': '*',
            'destination': '192.168.123.1/24',
            'protocol': 'any',
            'debug_logging': True,
            'pci_logging': False,
        }]

        # Get Subnets and used IPs for use in setting VM IPs
        data.used = [subnet['address_range'].split('/')[0] for subnet in data.subnets]

        # Add VM
        subnet = random.choice(data.subnets)
        ip = data.choose_ip(subnet['address_range'])
        data.used.append(ip)
        image = random.choice(unix_images) if unix else random.choice(windows_images)
        data.vms.append({
            'image_id': f'{image["id"]}',
            'name': f'TestVM-{image["display_name"].replace(" ", "-")}',
            'ram': ram,
            'gateway_subnet': subnet['address_range'],
            'ip_addresses': [
                {
                    'address': ip,
                    'nat': True,
                },
            ],
            'dns': '8.8.8.8,8.8.4.4',
            'cpu': cores,
            'storage_type_id': storage_type_id,
            'storages': [
                {
                    'primary': True,
                    'name': 'TestHD',
                    'gb': storage,
                },
            ],
        })

        return data
